devito/arguments.py

```diff
@@ -158,198 +145,398 @@ class PtrArgument(Argument):
+                # We've made sure the list isn't empty. Names should be unique so it
+                # should have exactly one entry
+                assert(len(orig_param_l) == 1)
+                orig_param = orig_param_l[0]
+                # Pull out the children and add them to kwargs
+                for orig_child, new_child in zip(orig_param.children, v.children):
+                    new_params[orig_child.name] = new_child
+        kwargs.update(new_params)
+        return kwargs
+
+    def _dle_arguments(self, dim_sizes):
+        # Add user-provided block sizes, if any
+        dle_arguments = OrderedDict()
+        autotune = True
+        for i in self.dle_arguments:
+            dim_size = dim_sizes.get(i.original_dim.name, None)
+            if dim_size is None:
+                raise InvalidArgument('Unable to derive size of dimension %s from '
+                                      'defaults. Please provide an explicit '
+                                      'value.' % i.original_dim.name)
+            if i.value:
+                try:
+                    dle_arguments[i.argument.name] = i.value(dim_size)
+                except TypeError:
+                    dle_arguments[i.argument.name] = i.value
+                    autotune = False
+            else:
+                dle_arguments[i.argument.name] = dim_size
+        return dle_arguments, autotune
+
+    def _derive_values(self, kwargs):
+        """ Populate values for all the arguments. The values provided in kwargs will
+            be used wherever provided. The values for the rest of the arguments will be
+            derived from the ones provided. The default values for the tensors come from
+            the data property of the symbols used in the Operator.
         """
-        raise NotImplemented()
+        values = OrderedDict()
+
+        values = OrderedDict([(i, get_value(i, kwargs.pop(i.name, None), values))
+                              for i in self.arguments])
+
+        dimension_values = OrderedDict([(i, kwargs.pop(i.name, None))
+                                        for i in self.dimension_params])
+
+        # Make sure we've used all arguments passed
+        if len(kwargs) > 0:
+            raise InvalidArgument("Unknown arguments passed: " + ", ".join(kwargs.keys()))
+
+        # Derive values for other arguments
+        for i in self.arguments:
+            if values[i] is None:
+                known_values = OrderedDict(chain(values.items(),
+                                                 dimension_values.items()))
+                provided_values = [get_value(i, x, known_values)
+                                   for x in i.gets_value_from]
+                assert(len(provided_values) > 0)
+                values[i] = reduce(max, provided_values)
+
+        # Second pass to evaluate any Unevaluated dependencies from the first pass
+        for k, v in values.items():
```

Comment:
could be made more pythonic

Reply:
I can't see how. Note that I'm modifying a dict in place, not creating one.